src/modules/iam/iam.service.unit.spec.ts

```diff
@@ -0,0 +1,147 @@
+import { Test, TestingModule } from '@nestjs/testing';
+import { Logger } from '../logger/logger.service';
+import { IamService } from './iam.service';
+
+const MockLogger = {
+  log: jest.fn(),
+  setContext: jest.fn(),
+};
+
+const domainsService = {
+  changeOrgOwnership: jest.fn(),
+  createOrganization: jest.fn(),
+};
+
+jest.mock('iam-client-lib', () => ({
+  setCacheConfig: jest.fn(),
+  setChainConfig: jest.fn(),
+  initWithPrivateKeySigner: jest.fn().mockImplementation(() => {
+    return {
+      connectToCacheServer: jest.fn().mockImplementation(() => {
+        return {
+          domainsService,
+          connectToDidRegistry: jest.fn().mockImplementation(() => {
+            return {
+              claimsService: jest.fn(),
+            };
+          }),
+        };
+      }),
+    };
+  }),
+}));
+
+describe('IAM Service', () => {
+  let service: IamService;
+
+  beforeEach(async () => {
+    const module: TestingModule = await Test.createTestingModule({
+      providers: [
+        IamService,
+        {
+          provide: ConfigService,
+          useValue: {
+            get: jest.fn(),
+          },
+        },
+        {
+          provide: Logger,
+          useValue: MockLogger,
+        },
+      ],
+    }).compile();
+
+    service = module.get<IamService>(IamService);
+    await service.initializeIAM();
+  });
+
+  /**
+   * An account can only execute one transaction on a blockchain at a time.
+   * Therefore, IAM service methods with update blockchain state should be executed in sequence,
+   * even if initiation of the methods is done concurrently
+   */
+  describe('blockchain non-concurrency', () => {
+    let blockchainInUse: boolean;
+    const blockChainTxDurationMillis = 10;
+    const blockchainTx = () => {
+      expect(blockchainInUse).toBeFalsy();
+      blockchainInUse = true;
+      return new Promise((resolve) => {
+        setTimeout(() => {
+          blockchainInUse = false;
+          resolve('');
+        }, blockChainTxDurationMillis);
+      });
+    };
+    const txFailedError = new Error('failed from timeout');
+    const failedBlockchainTx = () => {
+      expect(blockchainInUse).toBeFalsy();
+      blockchainInUse = true;
+      return new Promise((_, reject) => {
+        setTimeout(() => {
+          blockchainInUse = false;
+          reject(txFailedError);
+        }, blockChainTxDurationMillis);
+      });
+    };
+
+    beforeEach(() => {
+      blockchainInUse = false;
+      domainsService.changeOrgOwnership.mockImplementation(blockchainTx);
+      domainsService.createOrganization.mockImplementation(blockchainTx);
+    });
+
+    it(`should execute concurrent calls to changeOrganizationOwnership in sequence`, async () => {
+      const blockchainOperations = ['1', '2'].map((newOwner) => {
+        return service.changeOrgOwnership({
+          namespace: '',
+          newOwner,
+        });
+      });
+      await Promise.all(blockchainOperations);
```

Comment:
It seems strange to me that test is passing. Both blockchain operations expect that `blockchainInUse` isn't set and then set it. So they should not be able to run concurrently. Do you know how both operations has managed to resolve?

Reply:
> So they should not be able to run concurrently. 

Yes, that's point of these tests and code change, that they **don't** run concurrently. Like they can be triggered/queued concurrently (a new request can be queued while waiting on the I/O to complete) but two I/O operations that involve the blockchain for the same account shouldn't be executed concurrently only one can succeed.

Reply:
Ah, thanks. I understood it now